cluster/src/main/scala/org/apache/spark/scheduler/cluster/SnappyCoarseGrainedSchedulerBackend.scala

```diff
@@ -76,11 +76,12 @@ class BlockManagerIdListener(sc: SparkContext)
     val executorCores = msg.executorInfo.totalCores
     val profile = Misc.getMemStore.getDistributionAdvisor
         .getProfile(msg.executorId)
-    val numProcessors = if (profile != null) profile.getNumProcessors
-    else executorCores
+    val (numProcessors, usableHeap) =
+      if (profile != null) profile.getNumProcessors -> profile.getUsableHeap
+      else executorCores -> 0L
     SnappyContext.getBlockId(msg.executorId) match {
       case None => SnappyContext.addBlockId(msg.executorId,
-        new BlockAndExecutorId(null, executorCores, numProcessors))
+        new BlockAndExecutorId(null, executorCores, numProcessors, usableHeap))
```

Comment:
Why are we tracking usable heap on the lead node/driver side? I mean if we have failed task handling then anyways it will adjust in few iterations. May be this is more smart but this looping we are doing for not letting the servers die. Perhaps not keeping it may also be fine. As usable heap may change or is this just coming through file exchange and so it is the max usable configured heap being tracked and calculations being done based on that and not dynamic?

Reply:
Dynamic adjustment is "last line of defence". Initial setup calculates the spark.task.cpus assuming 128MB block size which is default for hadoop/parquet/ORC: 128MB * total-sparkCores / totalHeap. It uses max of that value and sparkCores/physicalCores. So the usableHeap per server is used to calculate totalHeap available across the cluster.